src/Message/AbstractRequest.php

```diff
@@ -333,24 +343,40 @@ function ($event) {
             }
         );
 
+        /** @var RequestInterface $httpRequest */
         $httpRequest = $this->httpClient->createRequest(
             $this->getHttpMethod(),
             $this->getEndpoint(),
             null,
             $data
         );
 
-        /**
-         * @var \Guzzle\Http\Message\RequestInterface
-         */
-        $httpRequest = $httpRequest
-                      ->setHeader(
-                          'Authorization',
-                          'Basic ' . base64_encode(($this->getUsername() ?: '') . ':' . ($this->getPassword() ?: ''))
-                      )
-                      ->setHeader('Content-Type', 'application/xml')
-                      ->setHeader('bluesnap-version', self::API_VERSION);
-        $httpResponse = $httpRequest->send();
+        $httpRequest
+            ->setHeader(
+                'Authorization',
+                'Basic ' . base64_encode(($this->getUsername() ?: '') . ':' . ($this->getPassword() ?: ''))
+            )
+            ->setHeader('Content-Type', 'application/xml')
+            ->setHeader('bluesnap-version', self::API_VERSION);
+
+        $httpResponse = null;
+        try {
+            // Fire a request event before sending request.
+            $eventDispatcher->dispatch(
```

Comment:
in omnipay-vindicia, you did this outside of the try, which i think makes a bit more sense. if dispatch() failed, i think we wouldn't want it to be caught and then dispatch an error, we'd want it to throw an exception.

Reply:
Sure, I'll move it outside.